scikeras/_saving_utils.py

```diff
@@ -0,0 +1,155 @@
+import os
+import zipfile
+
+from io import BytesIO
+from types import MethodType
+from uuid import uuid4 as uuid
+
+import numpy as np
+
+from tensorflow import io as tf_io
+from tensorflow import keras
+from tensorflow.keras.models import load_model
+from tensorflow.python.framework.errors_impl import NotFoundError
+from tensorflow.python.saved_model import loader_impl
+
+
+ram_prefix = "ram://"
+
+
+def new_parse_saved_model_with_debug_info(export_dir):
+    """Reads the savedmodel as well as the graph debug info.
+
+    Args:
+    export_dir: Directory containing the SavedModel and GraphDebugInfo files.
+
+    Returns:
+    `SavedModel` and `GraphDebugInfo` protocol buffers.
+
+    Raises:
+    IOError: If the saved model file does not exist, or cannot be successfully
+    parsed. Missing graph debug info file is fine.
+    """
+    saved_model = loader_impl._parse_saved_model(export_dir)
+
+    debug_info_path = os.path.join(
+        loader_impl.saved_model_utils.get_debug_dir(export_dir),
+        loader_impl.constants.DEBUG_INFO_FILENAME_PB,
+    )
+    debug_info = loader_impl.graph_debug_info_pb2.GraphDebugInfo()
+    if loader_impl.file_io.file_exists(debug_info_path):
+        with loader_impl.file_io.FileIO(debug_info_path, "rb") as debug_file:
+            try:
+                debug_info.ParseFromString(debug_file.read())
+            except loader_impl.message.DecodeError as e:
+                raise IOError("Cannot parse file %s: %s." % (debug_info_path, str(e)))
+            except NotFoundError:
+                # No debug info, loader_impl.file_io.file_exists is broken for ram://
+                pass
+    return (saved_model, debug_info)
+
+
+loader_impl.parse_saved_model_with_debug_info = new_parse_saved_model_with_debug_info
+
+
+def _temp_create_all_weights(self, var_list):
+    """A hack to restore weights in optimizers that use slots.
+
+    See https://tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer#slots_2
+    """
+    self._create_all_weights_orig(var_list)
+    try:
+        self.set_weights(self._restored_weights)
+    except ValueError:
+        # Weights don't match, eg. when optimizer was pickled before any training
+        pass
+    del self._restored_weights
+    self._create_all_weights = self._create_all_weights_orig
+
+
+def _restore_optimizer_weights(optimizer, weights) -> None:
+    optimizer._restored_weights = weights
+    optimizer._create_all_weights_orig = optimizer._create_all_weights
+    optimizer._create_all_weights = MethodType(_temp_create_all_weights, optimizer)
```

Comment:
Why is `MethodType` required? Does it have the same affect as [functools.partial][1]?

[1]:https://docs.python.org/3.8/library/functools.html#functools.partial

Reply:
MethodType binds the method to the live object: https://stackoverflow.com/a/37455782/6582418

Reply:
Got it. It doesn't serialize (/pickle/curry) the optimizer, it passes the memory address.

I think a comment would help clarify (which will almost certainly save time in the future).

Reply:
Added a comment.